src/models/rule.rs

```diff
@@ -203,6 +203,8 @@ impl InstantiatedRule {
       .iter()
       .filter_map(|h| substitutions.get(h).map(|s| (h.to_string(), s.to_string())))
       .collect();
+    // Since filter_map (above) discards any element of `rules.holes()` for which there isn't a valid substitution,
+    // Checking that the lengths match is enough to verify all holes have a matching substitution.
```

Comment:
```suggestion
    // checking that the lengths match is enough to verify all holes have a matching substitution.
```

No need for upper case after a comma.